packages/server-wallet/src/handlers/join-channel.ts

```diff
@@ -0,0 +1,62 @@
+import {Either, left, right, chain, map} from 'fp-ts/lib/Either';
+import {StateVariables} from '@statechannels/wallet-core';
+import {pipe} from 'fp-ts/lib/function';
+import {ChannelId} from '@statechannels/client-api-schema';
+
+import {SignState} from '../protocols/actions';
+import {ChannelState} from '../protocols/state';
+
+type HandlerResult = Either<JoinChannelError, SignState>;
+type StepResult = Either<JoinChannelError, ChannelState>;
+export interface JoinChannelHandlerParams {
+  channelId: ChannelId;
+}
+
+export enum Errors {
+  channelNotFound = 'channel not found',
+  invalidTurnNum = 'latest state must be turn 0',
+  alreadySignedByMe = 'already signed prefund setup',
+}
+
+export class JoinChannelError extends Error {
+  readonly type = 'JoinChannelError';
+  constructor(reason: Errors, public readonly data: any = undefined) {
+    super(reason);
+  }
+}
+
+// The helper functions should be factored out, tested, and reusable
+const hasStateSignedByMe = (cs: ChannelState): boolean => !!cs.latestSignedByMe;
+
+const ensureNotSignedByMe = (cs: ChannelState): StepResult =>
+  hasStateSignedByMe(cs) ? left(new JoinChannelError(Errors.alreadySignedByMe)) : right(cs);
+
+function ensureLatestStateIsPrefundSetup(cs: ChannelState): StepResult {
+  if (cs.latest.turnNum === 0) return right(cs);
+  return left(new JoinChannelError(Errors.invalidTurnNum));
+}
+
+function latest(cs: ChannelState): StateVariables {
+  return cs.latest;
+}
+// End helpers
+
+export function joinChannel(
+  args: JoinChannelHandlerParams,
+  channelState: ChannelState
+): HandlerResult {
+  // TODO: use Action creator from another branch
+  const signStateAction = (sv: StateVariables): SignState => ({
+    type: 'SignState',
+    channelId: args.channelId,
+    ...sv,
+  });
+
+  return pipe(
+    channelState,
+    ensureNotSignedByMe,
+    chain(ensureLatestStateIsPrefundSetup),
```

Comment:
Why are one of these functions wrapped in `chain` and the other not? They seem to have the same signature

Reply:
While true, they do not return a compatible type as input. If I have `f: (a: T1)=> T2` and `g: (a: T1) => T2` then I cannot compose them (unless `T2` extends `T1`)

`chain` changes a function to accept an `Either`, and either
- propagate the error, if given a `left`
- call `ensureLatestStateIsPrefundSetup` on the value, if given a `right`